File: 2022/haskell/app/X3.hs
import Data.List (elemIndex, nub)
import Data.Maybe (fromJust)
import Data.List.Split (chunksOf)


main :: IO ()
main = do
  input <- lines <$> readFile "../inputs/3"

  -- part one
  let priority c = fromJust (elemIndex c (['a'..'z'] ++ ['A'..'Z'])) + 1

      halve str =
        let half = length str `div` 2
        in (take half str, drop half str)

  print $ sum [priority c | (l, r) <- map halve input, c <- nub l, c `elem` r]

  -- part two
  print $ sum [priority z | [a, b, c] <- chunksOf 3 input, z <- nub a, z `elem` b && z `elem` c]

File: 2022/rust/src/x3.rs
use itertools::Itertools;
use std::fs;
use std::collections::HashSet;

pub fn main() {
    let input = fs::read_to_string("../inputs/3").expect("no file?");
    let data = input.trim().split("\n");

    // part one
    fn halve(s: &str) -> (String, String) {
        (s[0..s.len() / 2].to_string(), s[s.len() / 2..].to_string())
    }

    fn intersection((fst, snd): (String, String)) -> Vec<char> {
        let mut a: HashSet<char> = HashSet::from_iter(fst.chars());
        let b: HashSet<char> = HashSet::from_iter(snd.chars());
        a.retain(|&k| b.contains(&k));
        a.into_iter().collect()
    }

    fn priority(c: &char) -> i32 {
        ('a'..='z').chain('A'..='Z').position(|l| l == *c).unwrap() as i32 + 1
    }

    println!("{:?}", data.clone()
                         .map(halve)
                         .map(intersection)
                         .concat()
                         .iter()
                         .map(priority)
                         .sum::<i32>());

    // part two
    println!("{:?}", data.collect::<Vec<_>>()
                         .chunks(3)
                         // intersect three:
                         .map(|xs| intersection(
                                     (intersection((xs[0].to_string(),
                                                    xs[1].to_string()))
                                        .into_iter()
                                        .collect(),
                                     xs[2].to_string())))
                         .concat()
                         .iter()
                         .map(priority)
                         .sum::<i32>());

}
